File: backends/node/shared/utilities/src/string-extensions.ts
/**
 * String utility functions.
 * Mirrors D2.Shared.Utilities.Extensions.StringExtensions in .NET.
 */

const WHITESPACE_RE = /\s+/g;
const EMAIL_RE = /^[^@\s]+@[^@\s]+\.[^@\s]+$/;
const NON_DIGITS_RE = /[^\d]/g;
/**
 * Matches characters NOT allowed in display names.
 * Allowed: letters (any Unicode script), digits, spaces, hyphens,
 * apostrophes, periods, commas.
 *
 * Use with `str.replace(DISPLAY_NAME_INVALID_RE, "")` to strip, or
 * `DISPLAY_NAME_INVALID_RE.test(str)` to validate in form inputs.
 *
 * Exported for frontend form input masking (shared rules, no drift).
 */
export const DISPLAY_NAME_INVALID_RE = /[^\p{L}\p{N}\s\-'.,]/gu;

/**
 * Cleans a string by trimming leading/trailing whitespace and collapsing
 * duplicate whitespace into a single space.
 *
 * Returns `undefined` if the input is null, undefined, empty, or whitespace-only.
 */
export function cleanStr(str: string | null | undefined): string | undefined {
  const trimmed = str?.trim();
  if (!trimmed) return undefined;
  return trimmed.replace(WHITESPACE_RE, " ");
}

/**
 * Cleans a display name by stripping dangerous/unreasonable characters
 * (HTML tags, markdown syntax, brackets, quotes, backticks, etc.),
 * then trims whitespace and collapses duplicates.
 *
 * Allowed: letters (any Unicode script), digits, spaces, hyphens,
 * apostrophes, periods, commas.
 *
 * Returns `undefined` if the result is empty after cleaning.
 *
 * @example
 * cleanDisplayStr("John O'Brien-Smith") // "John O'Brien-Smith"
 * cleanDisplayStr("<script>alert(1)</script>") // "scriptalert1script"
 * cleanDisplayStr("**bold** [link](url)") // "bold linkurl"
 * cleanDisplayStr("Dr. José María") // "Dr. José María"
 */
export function cleanDisplayStr(name: string | null | undefined): string | undefined {
  if (name == null) return undefined;
  const stripped = name.replace(DISPLAY_NAME_INVALID_RE, "");
  return cleanStr(stripped);
}

/**
 * Cleans, normalizes, and validates the basic structure of an email address.
 *
 * @returns A normalized, cleaned, lowercased email address.
 * @throws {Error} If the email is null, empty, whitespace, or not in a valid format.
 */
export function cleanAndValidateEmail(email: string | null | undefined): string {
  const cleaned = cleanStr(email)?.toLowerCase();
  if (!cleaned || !EMAIL_RE.test(cleaned)) {
    throw new Error("Invalid email address format.");
  }
  return cleaned;
}

/**
 * Cleans and normalizes a phone number by removing all non-digit characters
 * and validating its length.
 *
 * @returns A normalized phone number (E.164 format — digits only, no leading "+").
 * @throws {Error} If the phone number is null, empty, or has fewer than 7 or more than 15 digits.
 */
export function cleanAndValidatePhoneNumber(phoneNumber: string | null | undefined): string {
  if (!phoneNumber?.trim()) {
    throw new Error("Phone number cannot be null or empty.");
  }

  const cleaned = phoneNumber.replace(NON_DIGITS_RE, "");

  if (!cleaned) {
    throw new Error("Invalid phone number format.");
  }

  if (cleaned.length < 7 || cleaned.length > 15) {
    throw new Error("Phone number must be between 7 and 15 digits in length.");
  }

  return cleaned;
}

/**
 * Returns the trimmed string if non-empty, or `undefined` if the value is
 * null, undefined, empty, or whitespace-only.
 *
 * Use at boundaries (user input, DB rows, proto mapping) to convert
 * empty strings to `undefined` before they propagate as "valid" data.
 *
 * Unlike {@link cleanStr}, this does NOT collapse internal whitespace —
 * it only trims and checks for emptiness.
 */
export function truthyOrUndefined(value: string | null | undefined): string | undefined {
  if (value == null) return undefined;
  const trimmed = value.trim();
  return trimmed.length > 0 ? trimmed : undefined;
}

/**
 * Generates a normalized string for hashing by cleaning and lowercasing each part,
 * then joining them with a pipe ("|") character.
 *
 * @example
 * getNormalizedStrForHashing([' Test One ', '   ', 'TEST3'])
 * // Returns: 'test one||test3'
 */
export function getNormalizedStrForHashing(parts: (string | null | undefined)[]): string {
  return parts.map((p) => cleanStr(p)?.toLowerCase() ?? "").join("|");
}

File: backends/dotnet/shared/Utilities/Extensions/StringExtensions.cs
// -----------------------------------------------------------------------
// <copyright file="StringExtensions.cs" company="DCSV">
// Copyright (c) DCSV. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------

namespace D2.Shared.Utilities.Extensions;

using System.Text.RegularExpressions;

/// <summary>
/// Extension methods for strings.
/// </summary>
public static partial class StringExtensions
{
    /// <summary>
    /// Extension methods for strings.
    /// </summary>
    ///
    /// <param name="str">
    /// The string.
    /// </param>
    extension(string? str)
    {
        /// <summary>
        /// Checks if a string is "truthy" (not null, not empty, not whitespace).
        /// </summary>
        ///
        /// <returns>
        /// Whether the string is truthy.
        /// </returns>
        public bool Truthy() => !str.Falsey();

        /// <summary>
        /// Checks if a string is "falsey" (null, empty, or whitespace).
        /// </summary>
        ///
        /// <returns>
        /// Whether the string is falsey.
        /// </returns>
        public bool Falsey() => string.IsNullOrWhiteSpace(str);

        /// <summary>
        /// Returns null if the string is null, empty, or whitespace-only; otherwise
        /// returns the trimmed string. Use at boundaries (DB rows, proto mapping,
        /// user input) to convert empty strings to null before they propagate.
        /// </summary>
        ///
        /// <returns>
        /// The trimmed string, or null if falsey.
        /// </returns>
        public string? ToNullIfEmpty()
        {
            if (str.Falsey())
            {
                return null;
            }

            return str!.Trim();
        }

        /// <summary>
        /// Cleans a string by trimming leading/trailing whitespace and replacing duplicate whitespace.
        /// </summary>
        ///
        /// <returns>
        /// The string, cleaned or null (if empty or null).
        /// </returns>
        public string? CleanStr()
        {
            var trimmed = str?.Trim();

            if (trimmed.Falsey())
            {
                return null;
            }

            return WhitespaceRegex().Replace(trimmed!, " ");
        }

        /// <summary>
        /// Cleans a display name by stripping dangerous/unreasonable characters
        /// (HTML tags, markdown syntax, brackets, quotes, backticks, etc.),
        /// then trims whitespace and collapses duplicates.
        /// <para>
        /// Allowed: letters (any Unicode script), digits, spaces, hyphens,
        /// apostrophes, periods, commas.
        /// </para>
        /// <para>
        /// Mirrors <c>cleanDisplayStr()</c> in <c>@d2/utilities</c> (TypeScript).
        /// </para>
        /// </summary>
        ///
        /// <returns>
        /// The cleaned display name, or null if the result is empty after cleaning.
        /// </returns>
        public string? CleanDisplayStr()
        {
            if (str.Falsey())
            {
                return null;
            }

            var stripped = DisplayNameInvalidRegex().Replace(str!, string.Empty);
            return stripped.CleanStr();
        }
    }

    /// <summary>
    /// Extension methods for email addresses.
    /// </summary>
    ///
    /// <param name="email">
    /// The email address.
    /// </param>
    extension(string? email)
    {
        /// <summary>
        /// Cleans, normalizes and [validates the basic structure of] an email address.
        /// </summary>
        ///
        /// <returns>
        /// A string containing a normalized, cleaned email address.
        /// </returns>
        ///
        /// <exception cref="ArgumentException">
        /// Thrown if the email is null, empty, whitespace, or not in a valid format.
        /// </exception>
        public string CleanAndValidateEmail()
        {
            var cleaned = email.CleanStr()?.ToLowerInvariant();
            if (cleaned.Falsey() || !EmailRegex().IsMatch(cleaned!))
            {
                throw new ArgumentException("Invalid email address format.", nameof(email));
            }

            return cleaned!;
        }
    }

    /// <summary>
    /// Extension methods for phone numbers.
    /// </summary>
    ///
    /// <param name="phoneNumber">
    /// The phone number.
    /// </param>
    extension(string? phoneNumber)
    {
        /// <summary>
        /// Cleans and normalizes a phone number by removing all non-digit characters and validating
        /// its length.
        /// </summary>
        ///
        /// <returns>
        /// A string containing a normalized, cleaned phone number (E.164 format - digits only - no
        /// leading "+").
        /// </returns>
        ///
        /// <exception cref="ArgumentException">
        /// Thrown if the phone number is null, empty, less than 7 or greater than 15 digits or in
        /// an invalid format.
        /// </exception>
        public string CleanAndValidatePhoneNumber()
        {
            if (phoneNumber.Falsey())
            {
                throw new ArgumentException(
                    "Phone number cannot be null or empty.", nameof(phoneNumber));
            }

            var cleaned = NonDigitsRegex().Replace(phoneNumber!, string.Empty);

            if (cleaned.Falsey())
            {
                throw new ArgumentException("Invalid phone number format.", nameof(phoneNumber));
            }

            if (cleaned.Length is < 7 or > 15)
            {
                throw new ArgumentException(
                    "Phone number must be between 7 and 15 digits in length.", nameof(phoneNumber));
            }

            return cleaned;
        }
    }

    /// <summary>
    /// Extension methods for string arrays used in hash normalization.
    /// </summary>
    ///
    /// <param name="parts">
    /// The individual values you would like to normalize for hashing.
    /// </param>
    extension(string?[] parts)
    {
        /// <summary>
        /// Generates a normalized string for hashing by cleaning and lowercasing each part, then
        /// joining them with a pipe ("|") character.
        /// </summary>
        ///
        /// <returns>
        /// A normalized string suitable for hashing.
        /// </returns>
        ///
        /// <example>
        /// If you enter the values " Test One ", "   ", "TEST3", the resulting string will be
        /// "testone||test3".
        /// </example>
        public string GetNormalizedStrForHashing()
        {
            return string.Join("|", parts.Select(x => CleanStr(x)?.ToLowerInvariant() ?? string.Empty));
        }
    }

    /// <summary>
    /// A regular expression that matches a basic email format.
    /// </summary>
    ///
    /// <returns>
    /// A regex match if the string matches a basic email format.
    /// </returns>
    [GeneratedRegex(@"^[^@\s]+@[^@\s]+\.[^@\s]+$", RegexOptions.None, matchTimeoutMilliseconds: 250)]
    private static partial Regex EmailRegex();

    /// <summary>
    /// A regular expression that matches one or more whitespace characters.
    /// </summary>
    ///
    /// <returns>
    /// A regex match if whitespace is atomically detected.
    /// </returns>
    [GeneratedRegex(@"\s+", RegexOptions.None, matchTimeoutMilliseconds: 100)]
    private static partial Regex WhitespaceRegex();

    /// <summary>
    /// Matches characters NOT allowed in display names.
    /// Allowed: letters (any Unicode script), digits, spaces, hyphens, apostrophes, periods, commas.
    /// Mirrors <c>DISPLAY_NAME_INVALID_RE</c> in <c>@d2/utilities</c> (TypeScript).
    /// </summary>
    [GeneratedRegex(@"[^\p{L}\p{N}\s\-'.,]", RegexOptions.None, matchTimeoutMilliseconds: 100)]
    private static partial Regex DisplayNameInvalidRegex();

    /// <summary>
    /// A regular expression that matches anything that is not a digit.
    /// </summary>
    ///
    /// <returns>
    /// A regex match if non-digit characters are detected.
    /// </returns>
    [GeneratedRegex(@"[^\d]", RegexOptions.None, matchTimeoutMilliseconds: 100)]
    private static partial Regex NonDigitsRegex();
}
